Replace asyncio.wait with asyncio.gather since wait ignores exceptions

ziv1234 opened the pull request:
fix for test_entity_platform so it expects the exception

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Not a breaking change

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fix for the helpers/test_entity_platform exceptions
Replaced asyncio.wait with asyncio.gather since wait doesn't propagate the exceptions to the waiting function. Once there, a duplicate unique_id triggers an exception that the test now catches.
Not my code, but just a suggestion

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ziv1234 committed:
replace asyncio.wait with asyncio.gather since wait ignores exceptions
fix for test_entity_platform so it expects the exception

balloob commented on homeassistant/helpers/entity_component.py:
```diff
@@ -128,7 +128,7 @@ def setup(self, config: ConfigType) -> None:
 
         if tasks:
-            await asyncio.wait(tasks)
+            await asyncio.gather(*tasks)
```
We shouldn't want this, because we don't want setup of 1 platform cancel the others. The same with the change for reset. We still want to log them.

balloob replied:
In all proposed changes here, it's ok to just log the errors and continue.

ziv1234 replied:
that makes sense, but brings up a few questions:
1. If we don't want this specific exception to be caught, perhaps instead of throwing the exception, just logging? Is there anywhere that we do catch it and do something with it?
2. If my change passed all the unit tests, but there is functionality that it breaks (we don't want setup of 1 platform cancel the others), perhaps we should add a unit tests that tests this scenario. I don't know enough about the startup process of hass to create this, but sounds like a simple test to prevent others from making my mistake :)
3. In any case, even if we cancel this exception, probably better to change the wait(tasks) to gather(*tasks) so the exceptions don't disappear and catch them in the calling method. I tried looking and it seems like _async_set_up_integrations in bootstrap.py may be the right place, but not sure

balloob replied:
1. We should log it in this method where we change from wait to gather. Gather can return exceptions.
2. Yeah that's bad. I'll add one.
3. Yeah

ziv1234 replied:
ok. just added the changes to _async_set_up_integrations 
I assume CORE_INTEGRATIONS don't need to be checked as if they fail, we fail anyway, right?

ziv1234 commented:
also, the same issue exists with async_add_entities with duplicate IDs. an exception gets thrown and never gets caught. This is the thing that fails the mqtt tests.
again, you guys know what the right behavior for the system, but if we don't want to propagate the exception, probably better to replace the exception with an error log.
Let me know what behavior is best and i can make the change.

ziv1234 committed:
changed to log on failed domains

ziv1234 committed:
discovered that this fix actually removes other uncaught exceptions

balloob commented:
Was actually just looking at this to add the tests (and got distracted obviously), but here is a quick update:

`EntityPlatform` should not raise HomeAssistantError on malformed or duplicate entity ID but instead should print an error and ignore entity.

ziv1234 committed:
fix in the list of ignored exceptions

ziv1234 commented:
> Was actually just looking at this to add the tests (and got distracted obviously), but here is a quick update:
> 
> `EntityPlatform` should not raise HomeAssistantError on malformed or duplicate entity ID but instead should print an error and ignore entity.

I agree. Will make the change tomorrow

ziv1234 committed:
replaced a few ignores on dyson tests that work locally but fail in the CI

ziv1234 committed:
two more tests that are failing on the CI and not locally

ziv1234 committed:
removed assertion on multiple entries with same unique_id - replaced with log and return
reverted test_entity_platform to its original since now there is no exception thrown

ziv1234 commented:
ok. made the change in entity_platform so it doesn't throw exception on duplicate unique_id and simply logs and returns.
I edited the list of ignored_exceptions, but the problem is that there are several tests that fail in the CI and succeed locally. Do you know why this can happen?
Hopefully I finally got the list right, but concerned about tests passing locally and not the CI as I try to not push a commit that fails tests...

MartinHjelmare commented on homeassistant/bootstrap.py:
```diff
@@ -325,15 +325,38 @@ def _get_domains(hass: core.HomeAssistant, config: Dict[str, Any]) -> Set[str]:
+                _LOGGER.error(
+                    "Error setting up integration %s - received exception %s",
+                    domain,
+                    str(futures[domain].exception()),
```
If this is an exception instance, it has a magic method `__str__` and we don't need to copy to string.

ziv1234 replied:
fixed

MartinHjelmare commented on homeassistant/helpers/entity_platform.py:
```diff
@@ -433,8 +433,8 @@ def add_entities(
-                msg += f". Platform {self.platform_name} does not generate unique IDs"
-            raise HomeAssistantError(msg)
+                msg += f". Platform {self.platform_name} does not generate unique IDs - ignoring entity"
+            return
```
We need to log the error if we remove the raise.

I'm not sure we want to remove the raise here though.

ziv1234 replied:
I agree that we need the log. we can either raise the exception or not, but if we do, it should be caught somewhere, otherwise why do we raise it.
Also, there are other exceptions there that have to be caught, so the catch in bootstrap.py is necessary anyway. Only question is what we do when there is a duplicate unique_id. You guys understand the architecture much better than me, so let me know what you prefer. In any case, it will be a minor change, but i need to know what the optimal expected behavior is.
Thanks

ziv1234 replied:
anyway, pushed a commit that adds the log
please tell me if you want the exception back. It won't really change the behavior since no one was catching the exception anyway so it will be caught in the component async_setup and all other entities will be added (if i understand the behavior correctly)

balloob replied:
I think that not raising is ok. The raising was only breaking the entity component, who would then ignore it. I guess we could consider raising an error and then having the entity component unloading that platform? That way we punish platforms harder if they violate the rules.

ziv1234 replied:
couldn't this introduce more instability that what it saves? i am no expert in hass, but unless we have many cases of components not returning unique_ids, you have to ask whether this is really helping.
out of curiosity, where would we unload the component? is that entity_component.py?

balloob replied:
We never unload components. We do reset components if we are reloading configuration. Platforms can be unloaded if the config entry is unloaded.

balloob replied:
Unloading a platform would mainly be a clear indicator to the user that the platform is broken and that the author should fix it. Let's stick to logging for now.

ziv1234 committed:
entered all the dyson tests. they all pass locally and probabilistically fail in the CI

ziv1234 committed:
removed unnecessary str() for exception

ziv1234 committed:
added log message for duplicate entity_id / unique_id

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -325,15 +325,38 @@ def _get_domains(hass: core.HomeAssistant, config: Dict[str, Any]) -> Set[str]:
+                )
+            else:
+                _LOGGER.error(
+                    "Error setting up integration %s - returned False", domain
```
The entity platform is already dealing with return value of `False`

ziv1234 replied:
ok, so just remove the log in that case?

balloob replied:
Yeah.

ziv1234 replied:
fixed

ziv1234 committed:
removed log in case of False return value

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -325,15 +325,29 @@ def _get_domains(hass: core.HomeAssistant, config: Dict[str, Any]) -> Set[str]:
+        await asyncio.wait(futures.values())
+        errors = [domain for domain in domains if futures[domain].exception()]
+        for domain in errors:
+            _LOGGER.error(
```
For this one, we have set up `async_setup_component` in such a way that we already wrap every piece of code that we call from an integration with `try…catch`. So if any of this raises, it's truly unexpected. So we should make sure we pass the exception as `exc_info` so we can print a stack trace

ziv1234 replied:
ok. done

ziv1234 committed:
added exc_info

balloob commented on homeassistant/bootstrap.py:
```diff
@@ -335,10 +335,13 @@ def _get_domains(hass: core.HomeAssistant, config: Dict[str, Any]) -> Set[str]:
+                "Error setting up integration %s - received exception",
                 domain,
-                futures[domain].exception(),
+                **kwargs,  # type: ignore
```
Why do it like this and not do `exc_info=(type(exception), exception, exception.__traceback__)` ?

ziv1234 replied:
i just copied it from the default exception handler in core.py, but you are correct. i made the change. do you want me to change the one in core.py as well?

ziv1234 replied:
BTW, if i understand the python behavior correctly, the one in core.py can also be replaced by exc_info=True, as it happens during the context of the exception, but not sure as i am new to python

balloob replied:
Yes, if we're within a `catch` block, we should let the logger get it themselves.

ziv1234 committed:
change the use of exc_info

ziv1234 commented:
thanks! one thing that is still puzzling me is that there are some tests (at least some of dyson and the localfile test) that fail on the CI if i remove them from the ignore list and they run fine locally, both with pytest and tox.
Now that this has been merged, i can recreate in a separate "dummy" PR and open an issue, since it would be great if the local tests would simulate the CI as well as possible

balloob commented:
On CI we run things in parallel with `-n auto` and based on the number of cores, it means the order of tests can be different, which might impact it.

ziv1234 commented:
Ok. I find it failing quite predictably in the CI on specific tests and i wasn't able to get them to fail locally. I opened an issue (https://github.com/home-assistant/core/issues/33504) to try and understand because it happens consistently.

Anything else that is different? In general, the tests are usually running independent of each other as each one creates a new hass fixture, right?

MartinHjelmare commented:
CI VMs can be slow so time difference can be greater which can matter when playing with time, eg not patching time correctly and firing time changed events.
